### ISSUE TITLE
Support for SSD NVME USB Enclosures

### ISSUE THREAD

**Issue by alfredovilla01** [mynodebtc/mynode#554]
Hi all- has anyone tried to connect an SSD via an NVME USB enclosure?  

I bought a Sabrent EC-NVME enclosure to run a 1TB SSD.

The RPI4 mynode doesn't recognize the drive... I tried it on both USB 3 and 2 ports.  Any suggestions?

Specifically- I'm using:  https://www.amazon.com/Sabrent-EC-NVME-Aluminum-Enclosure-Nvme/dp/B07K4TZQ7D

This may be an enhancement request?

**Comment by Transisto** [comment]
That's weird, does it work well on a PC?

**Comment by tehelsper** [comment]
Possibly, I've never tried an NVMe drive / enclosure with a Raspi 4. I know the drivers can be pretty picky even with the enclosure for 2.5in SATA SSDs, so I would not expect to have much luck.

The enclosure suggested on mynodebtc.com/download has the best track record from what I've seen. Maybe an NVMe enclosure from UGREEN (same brand) could work? I've have not had much luck from Sabrent adapters. Some work fine, but then I'll see an identical one that just doesn't work.

**Comment by alfredovilla01** [comment]
Yes, works well on PC / Laptop / even Mac.  So go figure.  It's gotta be a driver.  I even went and bought the UGREEN, but it also didn't work.  Ugh. 

Well- I think it's an RPi's issue.  The USB devices they use may not be that universal after all for Linux.  But they all do very well with Windows.  The UGREEN I bought was:  

UGREEN M.2 NVMe SSD Enclosure USB C 3.1 Gen 2 to M-Key M&B-Key NVMe PCIe 10Gbps External Enclosure Thunderbolt 3 Compatible with MacBook Pro WD Samsung Toshiba 2230 2242 2260 2280 NVMe PCIe SSD

https://www.amazon.com/gp/product/B07NPFV21K/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1

I have a high end cable USB A to USB C... which I also thought might be the issue.

Well, it's hit or miss.  What I did instead is bought an 2.5in SSD and put it a clear case and connected just fine.

**Comment by alfredovilla01** [comment]
Hi all- just thought I'd close this out.  

Seems like RPI's are particular about their USB connectors and case controllers.  Most "cases" have a controller that won't "see" the real SSD you put inside (or let it recognize the drives built in controller).  I suppose there is not much to do about that.

Having said that... I thought I would go out and buy a SanDisk SSD 1TB to see if it would work.  It DOES!  Again, its probably because there is not special "controller" in between so that the RPi can "see" the SSD directly.

Anyway... I recommend this one:
https://www.amazon.com/SanDisk-1TB-Extreme-Portable-SDSSDE61-1T00-G25/dp/B08GTYFC37

Its overkill on its speed, but the price point, especially with pre-Xmas specials, make it very affordable... and on point with just about any 1TB SSD out there ;)

Good luck all!!

**Comment by Transisto** [comment]
Samsung T5 does the job too.

**Comment by tlindi** [comment]
Wouldn't this be solution, If it works (I'm just looking forward to purchase):

https://www.argon40.com/argon-one-m-2-case-for-raspberry-pi-4.html

**Comment by tehelsper** [comment]
Yeah, the Argon One M2 case works great for one of my devices. NVME drives should work fine, but some may require tweaks like listed above. The M2 didn't take anything extra (except running one command listed in the case's guide to install the fan driver).

**Comment by tlindi** [comment]
Just FYI Argon link has changed. And they have M.2 SATA and M.2 NVMe and they have many variations of chips with range of different compatibility levels with usb-storage / ufsb (?) kernel drivers having or not support for trim. Real mess :-/

**Comment by tlindi** [comment]
> Yeah, the Argon One M2 case works great for one of my devices. NVME drives should work fine, but some may require tweaks like listed above. The M2 didn't take anything extra (except running one command listed in the case's guide to install the fan driver).

I have now M.2 SATA and M.2 NVMe with updated firmware which fixed wrong product number issue. Addition to fan driver / settings app I like heartbeat pattern power led - which can be created with three lines.

But... could something like this be used to detect which case and which disk is used, and make quirks for devices that require such?

$ lsblk -S
```
NAME HCTL       TYPE VENDOR   MODEL                        REV SERIAL          TRAN
sda  0:0:0:0    disk Argon    Samsung SSD 860 EVO M.2 1TB 0    S415xxxxxxxxx 4T usb

$ USB_ID=$(lsusb | grep ASMedia | awk '{print $6}')
$ USB_VID=$(echo $USB_ID | awk -F: '{print $1}')
$ USB_PID=$(echo $USB_ID | awk -F: '{print $2}')
$ echo $USB_VID $USB_PID
174c 1156

$ lsusb -v -d $USB_ID | grep iManufacturer
Couldn't open device, some information will be missing
  iManufacturer           2 Argon

$ lsusb -v -d $USB_ID | grep iProduct
Couldn't open device, some information will be missing
  iProduct                3 Forty
```

Eg Argon One M.2 SATA needs quirks for TRIM to work. and with above this could be created automatically:
```
cat /etc/udev/rules.d/10-trim.rules
ACTION=="add|change", ATTRS{idVendor}=="174c", ATTRS{idProduct}=="1156", SUBSYSTEM=="scsi_disk", ATTR{provisioning_mode}="unmap"
```

Would udev be more general way, it would work also others than RPi too?
